src/Analyser/NodeScopeResolver.php

```diff
@@ -4522,8 +4524,22 @@ private function processArgs(
 			}
 
 			if ($assignByReference) {
+				if ($currentParameter === null) {
+					throw new ShouldNotHappenException();
+				}
+
 				$argValue = $arg->value;
 				if ($argValue instanceof Variable && is_string($argValue->name)) {
+
+					$paramOutTypes = $this->getParameterOutExtensionsTypes($callLike, $calleeReflection, $currentParameter, $scope);
+					if ($paramOutTypes !== []) {
+						if (!$byRefType instanceof MixedType) {
+							$paramOutTypes[] = $byRefType;
```

Comment:
I feel like extensions should override the byRefType, it should not be added in any case. Meaning all this logic should be in `else` below line 4521.

Reply:
I did it like that, because return-type extension work the same way. will adjust.

Reply:
It doesn't. it throws away the original return type when an extension returns a non-null value.

Reply:
oh sorry, my reply above was meant for this comment: https://github.com/phpstan/phpstan-src/pull/3083#discussion_r1627254141